src/Url.php

```diff
@@ -2,70 +2,298 @@
+     * @param string|null $url     URL.
+     * @param Url|null    $baseUrl Base URL.
+     * @throws FailedToParseUrl Exception when the URL or Base URL is malformed.
+     */
+    public function __construct($url = null, Url $baseUrl = null)
+    {
+        $parsedUrl = [];
+
+        if ($url !== null) {
+            $parsedUrl = parse_url($url);
+            if (false === $parsedUrl) {
+                throw FailedToParseUrl::forUrl($url);
+            }
+        }
+
+        if ($baseUrl) {
+            if (isset($parsedUrl[self::PATH]) && 0 !== strpos($parsedUrl[self::PATH], '/')) {
+                $parsedUrl[self::PATH] = $this->unrelativizePath("{$baseUrl->path}/{$parsedUrl[self::PATH]}");
+            }
+            $parsedUrl = array_merge($baseUrl->toArray(), $parsedUrl);
+        }
+
+        if ($parsedUrl) {
+            foreach ($parsedUrl as $key => $value) {
+                if (property_exists($this, $key)) {
+                    $this->$key = $value;
+                }
+            }
+        }
+    }
+
+    /**
+     * Eliminate relative segments (../ and ./) from a path.
+     *
+     * @param string $path Path with relative segments. This is not a URL, so no host and no query string.
+     * @return string Unrelativized path.
+     */
+    private function unrelativizePath($path)
+    {
+
+        // Eliminate current directory relative paths, like <foo/./bar/./baz.css> => <foo/bar/baz.css>.
+        do {
+            $path = preg_replace(
+                self::COLLAPSE_CURRENT_PATHS_REGEX_PATTERN,
+                '/',
+                $path,
+                -1,
+                $count
+            );
+        } while (0 !== $count);
+
+        // Collapse relative paths, like <foo/bar/../../baz.css> => <baz.css>.
+        do {
+            $path = preg_replace(
+                self::COLLAPSE_RELATIVE_PATHS_REGEX_PATTERN,
+                '',
+                $path,
+                1,
+                $count
+            );
+        } while (0 !== $count);
+
+        return $path;
+    }
+
+    /**
+     * Check whether the URL is a valid image source URL.
      *
-     * @param string $src Src string to validate.
      * @return bool Whether the src string is a valid image source URL.
      */
-    public static function isValidNonDataUrl($src)
+    public function isValidNonDataUrl()
     {
-        list($scheme, $host, $port, $user, $pass, $path, $query, $fragment) = array_values(
-            array_merge(
-                self::URL_DEFAULT_PARTS,
-                (array)parse_url($src)
-            )
-        );
-
-        if ($scheme === 'data') {
+        // Bail early on 'data:' assets.
+        if ($this->scheme === 'data') {
             return false;
         }
 
-        $userpass = $user;
+        // @TODO: This probably needs additional logic.
+
+        $parts = array_merge(self::URL_DEFAULT_PARTS, $this->toArray(true));
+
+        return (bool)filter_var(self::toString($parts), FILTER_VALIDATE_URL);
```

Comment:
We'll have to add `filter` to the required extensions.

Reply:
Done in ff5995e and https://github.com/ampproject/amp-wp/pull/6037/commits/178e68c813ffd2a11c49ccc7ad7b4e01b60986b9